building_sim_plugins/building_gazebo_plugins/src/crowd_simulator.cpp

```diff
@@ -182,125 +212,82 @@ void CrowdSimulatorPlugin::_update_internal_object(
-      gazebo::physics::ActorPtr actor_ptr =
-        boost::dynamic_pointer_cast<gazebo::physics::Actor>(model_ptr);
-      gazebo::physics::TrajectoryInfoPtr trajectory_info(new gazebo::physics::
-        TrajectoryInfo()); //matches the actor skeleton
-
-      crowd_simulator::ModelTypeDatabase::RecordPtr type_ptr =
-        _crowd_sim_interface->_model_type_db_ptr->get(obj_ptr->type_name);
-      trajectory_info->type = type_ptr->animation;
-      // set each keyframe duration as the sim_time_step
-      trajectory_info->duration = _crowd_sim_interface->get_sim_time_step();
-      actor_ptr->SetCustomTrajectory(trajectory_info);
-      actor_ptr->SetStatic(false);
-
-      //check actor has idle animation
-      for (auto idle_anim : _crowd_sim_interface->get_switch_anim_name())
+      for (size_t id = 0; id < crowd_sim_interface->_objects_count; id++)
       {
-        if (actor_ptr->SkeletonAnimations().find(idle_anim) !=
-          actor_ptr->SkeletonAnimations().end())
-        {
-          type_ptr->idle_animation = idle_anim;
-          break;
-        }
+        auto name = crowd_sim_interface->get_internal_agent(id);
+        gazebo::physics::ModelPtr model_ptr = _world->ModelByName(name);
+        gazebo::physics::ActorPtr actor_ptr =
+          boost::dynamic_pointer_cast<gazebo::physics::Actor>(model_ptr);
+        gazebo::physics::TrajectoryInfoPtr trajectory_info(new gazebo::physics::
+          TrajectoryInfo()); //matches the actor skeleton
+        trajectory_info->type = "idle";
+        actor_ptr->SetCustomTrajectory(trajectory_info);
+        actor_ptr->SetStatic(false);
       }
+      _initialized = true;
     }
-  }
-  _initialized = true;
-  RCLCPP_INFO(
-    _crowd_sim_interface->logger(),
-    "Gazebo models all loaded! Start simulating...");
-}
-
-//============================================
-bool CrowdSimulatorPlugin::_spawn_agents_in_world()
-{
-  //create model in world for each internal agents
-  _objects_count = _crowd_sim_interface->get_num_objects();
-  for (size_t id = 0; id < _objects_count; id++)
-  {
-    if (!_crowd_sim_interface->get_object_by_id(id)->is_external)
+    else
     {
-      auto object_ptr = _crowd_sim_interface->get_object_by_id(id);
-      assert(object_ptr);
-      auto type_ptr = _crowd_sim_interface->_model_type_db_ptr->get(
-        object_ptr->type_name);
-      assert(type_ptr);
-      if (!_create_model(object_ptr->model_name, type_ptr,
-        object_ptr->agent_ptr) )
+      for (size_t id = 0; id < crowd_sim_interface->_objects_count; ++id)
       {
-        RCLCPP_INFO(_crowd_sim_interface->logger(),
-          "Failed to insert model [" + object_ptr->model_name + "] in world");
-        return false;
+        ObjectPtr obj_ptr = crowd_sim_interface->get_object_by_id(id);
+        // spawned agents are not fully loaded
+        if (!_world->ModelByName(obj_ptr->model_name))
+        {
+          _initialized = false;
+          return;
+        }
```

Comment:
I keep this statement and associated checks from Fred's. The spawning comes in 2 parts:

1. models in sdf loaded into ignition/gazebo simulation world
2. human in resource_config files read by menge and then again read out from menge to get associated data.

We are dealing with check on (2) here in case the resource config or menge mess up, not the (1).

Reply:
Did you try to remove the check / add a breakpoint to see if it usually happens? I'm still not convinced since from what I understand

- You are iterating over all the models in menge, and those are initialized at loading time.
- You are checking them against the ones in the world, which are also initialized before the plugin update function is called.

What would be a typical case in which this statement is executed? I tried to run a few simulations on my machine and it never happened.

Reply:
I see it an assertion on the human in scene file matching that of human in world file.

U are right. In all cases they are the same, hence no need for the check. In the event where Menge scene file has different configuration, most likely other segmentation fault error will happen even before this check.

I can remove this "assertion" if you are okay with that.